src/vulkan/transfer_image.cc

```diff
@@ -155,42 +163,53 @@ Result TransferImage::CreateVkImageView() {
   return {};
 }
 
-VkBufferImageCopy TransferImage::CreateBufferImageCopy() {
+VkBufferImageCopy TransferImage::CreateBufferImageCopy(uint32_t mip_level) {
   VkBufferImageCopy copy_region = VkBufferImageCopy();
   copy_region.bufferOffset = 0;
   // Row length of 0 results in tight packing of rows, so the row stride
   // is the number of texels times the texel stride.
   copy_region.bufferRowLength = 0;
   copy_region.bufferImageHeight = 0;
   copy_region.imageSubresource = {
-      aspect_, /* aspectMask */
-      0,       /* mipLevel */
-      0,       /* baseArrayLayer */
-      1,       /* layerCount */
+      aspect_,   /* aspectMask */
+      mip_level, /* mipLevel */
+      0,         /* baseArrayLayer */
+      1,         /* layerCount */
   };
   copy_region.imageOffset = {0, 0, 0};
-  copy_region.imageExtent = {image_info_.extent.width,
-                             image_info_.extent.height, 1};
+  copy_region.imageExtent = {image_info_.extent.width >> mip_level,
+                             image_info_.extent.height >> mip_level, 1};
   return copy_region;
 }
 
 void TransferImage::CopyToHost(CommandBuffer* command_buffer) {
-  auto copy_region = CreateBufferImageCopy();
+  std::vector<VkBufferImageCopy> copy_regions;
+  uint32_t last_mip_level = used_mip_levels_ == VK_REMAINING_MIP_LEVELS
+                                ? mip_levels_
+                                : base_mip_level_ + used_mip_levels_;
+  for (uint32_t i = base_mip_level_; i < last_mip_level; i++)
+    copy_regions.push_back(CreateBufferImageCopy(i));
 
   device_->GetPtrs()->vkCmdCopyImageToBuffer(
       command_buffer->GetVkCommandBuffer(), image_,
-      VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, host_accessible_buffer_, 1,
-      &copy_region);
+      VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, host_accessible_buffer_,
+      static_cast<uint32_t>(copy_regions.size()), copy_regions.data());
 
   MemoryBarrier(command_buffer);
 }
 
 void TransferImage::CopyToDevice(CommandBuffer* command_buffer) {
-  auto copy_region = CreateBufferImageCopy();
+  std::vector<VkBufferImageCopy> copy_regions;
+  uint32_t last_mip_level = used_mip_levels_ == VK_REMAINING_MIP_LEVELS
+                                ? mip_levels_
+                                : base_mip_level_ + used_mip_levels_;
+  for (uint32_t i = base_mip_level_; i < last_mip_level; i++)
+    copy_regions.push_back(CreateBufferImageCopy(i));
 
   device_->GetPtrs()->vkCmdCopyBufferToImage(
       command_buffer->GetVkCommandBuffer(), host_accessible_buffer_, image_,
-      VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &copy_region);
+      VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL,
+      static_cast<uint32_t>(copy_regions.size()), copy_regions.data());
```

Comment:
How do these MIP levels end up in the buffer memory? We're copying N buffers into the same buffer memory, but we appear to be writing them as the same spot? Or does Vulkan do something behind the scenes to make sure they don't write on top of each other? Can you add some comments here on how this works?

Reply:
The copy region has the information which mip level is used. I believe implementations handle that how they want. That's why we need to query the buffer memory requirements from the driver too (with the number of mip maps).